### DISCUSSION TITLE
GraphQL Support

### DISCUSSION THREAD

**Discussion by chrisrowe** [solspace/craft-calendar#28] [Feature Requests]
Now this is baked into Craft I'm hesitant to use any plugin that doesn't support it.

**Comment by mildlygeeky** [comment]
Ditto - right now we're having to look at writing a custom controller that returns results in JSON to be consumed by a headless frontend, but really wish we could make a GQL call (perhaps requiring an end date / number of days to return to keep recurring events in check).

**Reply by gustavs-gutmanis** [reply]
We'll add support for GraphQL in the upcoming update :)

**Comment by kjmartens** [comment]
We've now added support for GraphQL in Calendar 3.3.0. :)

**Reply by saltymouse** [reply]
Any way to query custom fields in addition to basic event info?

**Reply by gustavs-gutmanis** [reply]
Not yet @saltymouse, but we'll implement it

**Reply by saltymouse** [reply]
Fabulous! I'll check back for updates—looking forward to this feature!

**Reply by jonathanmelville** [reply]
Is there an ETA on querying for custom fields?

**Reply by kjmartens** [reply]
This is available now in 3.3.1 :)

**Reply by jonathanmelville** [reply]
In production I'm getting this error, do you know what could cause it or how to resolve it? 

`[error][craft\errors\GqlException] craft\errors\GqlException: Tried to load an unregistered type "events_Event". This can indicate both a typo in the query or an issue with the schema used.`

**Reply by saltymouse** [reply]
I can't find it in the documentation yet, but for anyone else curious, using the `... on default_Event` inline fragment is how I've queried custom fields on my default Event type in v3.3.1. (The `featuredImage` field below is a custom Craft asset field attached to my default Event layout):

```graphql
query relatedEventsQuery ($eventId: [Int]) {
    solspace_calendar {
        calendar(id: 1) {
            events (id: $eventId) {
                id
                ...on default_Event {
                    featuredImage {
                        id
                        url @transform (width: 200, immediately: true)
                        width
                        height
                        title
                    }
                }
            }
        }
    }
}
```

**Comment by saltymouse** [comment]
The JSON Event interface before GraphQL (i.e. `event.json.twig`) has some useful properties that are used in the "Demo Templates" to build the pages, yet these are [unavailable via GraphQL](https://docs.solspace.com/craft/calendar/v3/developer/graphql.html#calendar-interface).

Missing properties I've noticed are:
- `multiDay`
- `url`
- `duration`

`multiDay` can be calculated fairly simply using an external library like [`isSameDay` from `date-fns`](https://date-fns.org/v2.22.1/docs/isSameDay), but it'd be nice to have the same properties included.

Without a `url` or `uri` or `slug` property to read from, it's difficult to generate a URL/URI link via JS to match what Craft's `slug` produces.

If these are currently possible, please let me know! Thanks!

**Reply by jonathanmelville** [reply]
@saltymouse I have a site under development where we're implementing Calendar in Vue w/ GQL. I ended up using a slug like `/events/_year/_month/_day/_id`, which would give you a URL similar to `/events/2021/07/03/188`, this way you access to all of the data that you need for displaying an individual event. This has worked well for me.

This site is still under development, but you can see an example of an individual recurring event here: https://the-wharf.vercel.app/events/2021/07/23/340

**Reply by saltymouse** [reply]
Thanks for the tip @jonathanmelville (and nice looking calendar so far!). I think I'll remove the title-based `slug` from the URL for now and stick with `/events/_year/_month/_day/_id` like you mentioned.

Some people like having a title-based `slug` in a URL for SEO purposes, so hopefully we can get that in GraphQL eventually.

**Reply by saltymouse** [reply]
@jonathanmelville I noticed your recurring events share the same ID in the URL since they originate from the same event, but you're able to apply the recurring event's future date in the URL as well.

However, when I set my Calendar's event slug to `event/{startDate|date('Y')}/{startDate|date('m')}/{startDate|date('d')}/{id}` this means all future recurring events end up using the _original_ event's `startDate` for the `_year/_month/_day` portion of the URL.

May I ask how you set this up allowing for dynamic date slugs with your repeating events? Custom route patterns instead of Calendar slug generation? Thanks!

---
Edit:
I was able to achieve the same effect by:
1. Removing the generated slug from the "Event URI Format" Calendar setting.
2. Adding a custom Craft route pattern with the following `(tokens)`: `event/(year)/(month)/(day)/(number)`
3. Extract the Event's date using the `(year)/(month)/(day)` URL params for rendering/presenting in the (recurring) event's date in the Event template (rather than what is retrieved from the `event.start` date (which was referring to the original event date for recurring events)).

**Reply by gustavs-gutmanis** [reply]
@saltymouse, @jonathanmelville
The upcoming version will have the `url`, `multiDay` and `duration` variables present in the GQL query.  
If you wish to try it out before the release, you can install the `dev-master` version of the plugin.

**Reply by jonathanmelville** [reply]
Hey @saltymouse, sorry I'm chiming in late on your question. Basically this app is "headless" and gets all it's data through GraphQL. So all the routing is handled on the frontend in Javascript.

**Reply by gustavs-gutmanis** [reply]
The new properties are now available in v3.3.2

**Comment by jonathanmelville** [comment]
@saltymouse No prob!

**Comment by gustavs-gutmanis** [comment]
Thanks for reporting this, we'll add support for these properties next week.

--  
Gustavs Gutmanis
